snuba/datasets/entity_subscriptions/entity_subscription.py

```diff
@@ -18,112 +17,49 @@ class InvalidSubscriptionError(Exception):
     pass
 
 
-class EntitySubscription(ABC):
-    def __init__(self, data_dict: Mapping[str, Any]) -> None:
-        ...
-
-    @abstractmethod
-    def get_entity_subscription_conditions_for_snql(
-        self, offset: Optional[int] = None
-    ) -> Sequence[Expression]:
-        """
-        Returns a list of extra conditions that are entity specific and required for the
-        snql subscriptions
-        """
-        raise NotImplementedError
-
-    @abstractmethod
-    def validate_query(self, query: Union[CompositeQuery[Entity], Query]) -> None:
-        """
-        Applies entity specific validations on query argument passed
-        """
-        raise NotImplementedError
-
-    @abstractmethod
-    def to_dict(self) -> Mapping[str, Any]:
-        raise NotImplementedError
+class EntitySubscription:
+    """
+    Used to provide extra functionality for entity specific subscriptions.
+    An instance of this class is set as an attribute of Entity.
+        Entity.entity_subscription: Optional[EntitySubscription]
+    """
 
+    def __init__(
+        self,
+        max_allowed_aggregations: int = 1,
+        disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"],
+    ) -> None:
+        self.max_allowed_aggregations = max_allowed_aggregations
+        self.disallowed_aggregations = disallowed_aggregations
 
-class EntitySubscriptionValidation:
-    MAX_ALLOWED_AGGREGATIONS: int = 1
-    disallowed_aggregations: Sequence[str] = ["groupby", "having", "orderby"]
+    def validate_query(self, query: Union[CompositeQuery[EntityDS], Query]) -> None:
+        # Import get_entity() when used, not at import time to avoid circular imports
```

Comment:
This import cycle is not caused by type checking, but from the nature of how entity factory is set up. Entity factory requires this `entity_subscription.py` module which has a specific function that looks up an entity. However, this is done in validation so it's guaranteed that entities are already created beforehand.

Reply:
This feels like our code is structured incorrectly. Do you think it would be difficult to move the validation functions out of the `EntitySubscription` class, and make them functions in the subscriptions module?

Reply:
I don't think it would be difficult to do that given how this validate query method is only called once in the subscriptions module. Is this something we want to commit to? Personally, I don't think there is a significant difference given how entity subscription is just a container for these methods used by the subscriptions module. However, if in the future we want to support entity specific validation on a query, then this might useful to keep?

Reply:
Agree this is a sign that the code is not structured correctly. If EntitySubscription is a property of an entity, it doesn't make sense for it to also require the entity. Even though you found a way around the circular import issue, this is a huge hack.